### PR TITLE
Component

### PR DESCRIPTION
**Pull request by lachrist**
Revamp of the component system and some minor fixes.

### CODE REVIEW

**Review by dividedmind — APPROVED** [review]
I think I have read almost all the substantial parts and understood about half of it. I definitely approve getting rid of explicit DI, I've never been a big fan :)

Looks good. I think. Sorry it took so long, but it's quite a handful!

File: build/component/util.mjs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  undefined,
+  Reflect: { getOwnPropertyDescriptor },
+} = globalThis;
+
+export const {
+  Object: {
+    hasOwn = (object, key) =>
+      getOwnPropertyDescriptor(object, key) !== undefined,
+  },
+} = globalThis;
```
[the remark this answers is not in the document]
Comment (lachrist) [reply build/component/util.mjs:11]: Not exactly, `hasOwn` will be assign to `Object.hasOwn` if it is not `undefined` which is faster than the polyfilled version.

File: build/component/index.mjs [hunk 2]
```diff
@@ -1,33 +1,160 @@
...
+const isDirectoryAsync = async (base, url) =>
+  (await statAsync(new URL(url, base))).isDirectory();
+
+const parseEnv = (content) => content.split(EOL).filter(isNotEmptyString);
+
+const loadEnvironmentsAsync = async (home, component, instance) => {
+  const envs = parseEnv(
+    await readFileAsync(new URL(`${component}/${instance}/.env`, home), "utf8"),
+  );
+  if (envs.every(isFirstOccurence)) {
```
[the remark this answers is not in the document]
Comment (lachrist) [reply build/component/index.mjs:42]: I actually follow your suggestion, I had not thought about using `Set` 🧇 . Not so much that it is faster but it requires less code. 52d2ab0c46949540b780597f093860b33ef30b27

File: build/component/util.mjs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+export const filterAsync = async (array, predicateAsync) => {
+  const filtered_array = [];
+  for (let index = 0; index < array.length; index += 1) {
```
[the remark this answers is not in the document]
Comment (lachrist) [reply build/component/util.mjs:15]: I need `index` to pass it to `predicateAsync`. I'm not sure that I actually use the index but `filterAsync` should resemble `Array.prototype.filter` as much as possible IMO.

File: components/loader.mjs [hunk 4]
```diff
@@ -0,0 +1,47 @@
+const {
+  Error,
+  undefined,
+  Reflect: { getOwnPropertyDescriptor, ownKeys },
+  Object: {
+    hasOwn = (object, key) =>
+      getOwnPropertyDescriptor(object, key) !== undefined,
+  },
```
[the remark this answers is not in the document]
Comment (lachrist) [reply components/loader.mjs:8]: Which `util`?
- `build/component/util.mjs` will not be included in production.
- `components/util/index.mjs` is circular

File: build/component/util.mjs (hunk 1 above)
[the remark this answers is not in the document]
Comment (dividedmind) [reply build/component/util.mjs:11]: Ah, ok, now I see what's happening here. I think :) Thanks

**Pull request merged**